File: Java_client/bim-portal-enhanced/src/main/java/com/pb40/bimportal/examples/DomainModelExportExample.java
package com.pb40.bimportal.examples;

import com.bimportal.client.model.SimpleDomainSpecificModelPublicDto;
import com.pb40.bimportal.client.BimPortalClientBuilder;
import com.pb40.bimportal.client.EnhancedBimPortalClient;
import com.pb40.bimportal.config.BimPortalConfig;
import com.pb40.bimportal.util.ExportUtils;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Domain model export examples for BIM Portal Java client.
 *
 * <p>Demonstrates domain-specific model export workflows in multiple formats including PDF,
 * OpenOffice, OKSTRA, LOIN-XML, and IDS with automatic content type detection.
 */
public class DomainModelExportExample {

  private static final Logger logger = LoggerFactory.getLogger(DomainModelExportExample.class);

  /**
   * Check if credentials are available for authentication.
   *
   * @return True if credentials are configured
   */
  private static boolean checkCredentials() {
    if (!BimPortalConfig.hasCredentials()) {
      System.out.println("============================================================");
      System.out.println("WARNING: Credentials not found in environment variables.");
      System.out.println(
          "Please set "
              + BimPortalConfig.USERNAME_ENV_VAR
              + " and "
              + BimPortalConfig.PASSWORD_ENV_VAR
              + " in .env file.");
      System.out.println("============================================================");
      return false;
    }
    return true;
  }

  /**
   * Detect content type from byte array and return appropriate file extension.
   *
   * @param content The byte array content
   * @param defaultExtension Default extension if detection fails
   * @return Detected file extension
   */
  private static String detectFileExtension(byte[] content, String defaultExtension) {
    if (content == null || content.length < 4) {
      return defaultExtension;
    }

    // Check for ZIP file signature (PK)
    if (content.length >= 2 && content[0] == 0x50 && content[1] == 0x4B) {
      logger.debug("Detected ZIP file signature");
      return "zip";
    }

    // Check for PDF signature (%PDF)
    if (content.length >= 4
        && content[0] == 0x25
        && content[1] == 0x50
        && content[2] == 0x44
        && content[3] == 0x46) {
      logger.debug("Detected PDF file signature");
      return "pdf";
    }

    // Check for OpenDocument format (ZIP-based, but starts differently)
    if (content.length >= 4 && content[0] == 0x50 && content[1] == 0x4B) {
      // Additional check for OpenDocument by looking for mimetype file
      // This is a more sophisticated check for ODT files
      String contentStr = new String(content, 0, Math.min(1024, content.length));
      if (contentStr.contains("mimetypeapplication/vnd.oasis.opendocument")) {
        logger.debug("Detected OpenDocument format");
        return "odt";
      }
      logger.debug("Detected ZIP-based format (possibly ODT or other)");
      return "zip";
    }

    // Check for XML content (IDS files are typically XML)
    if (content.length >= 5) {
      String start = new String(content, 0, Math.min(100, content.length));
      if (start.trim().startsWith("<?xml") || start.contains("<ids")) {
        logger.debug("Detected XML format");
        return "xml";
      }
    }

    logger.debug("Could not detect file type, using default: {}", defaultExtension);
    return defaultExtension;
  }

  /**
   * Export content with automatic file type detection.
   *
   * @param content The byte array content
   * @param baseFilename Base filename without extension
   * @param expectedExtension Expected file extension for logging
   * @return Optional path to saved file
   */
  private static Optional<Path> exportWithDetection(
      byte[] content, String baseFilename, String expectedExtension) {
    String detectedExtension = detectFileExtension(content, expectedExtension);
    String filename = baseFilename + "." + detectedExtension;

    if (!detectedExtension.equals(expectedExtension)) {
      logger.info(
          "Content type detection: expected '{}' but detected '{}' for {}",
          expectedExtension,
          detectedExtension,
          baseFilename);
    }

    return ExportUtils.saveExportFile(content, filename);
  }

  /**
   * Run domain model export examples.
   *
   * @param client Enhanced BIM Portal client
   */
  public static void runDomainModelExportExamples(EnhancedBimPortalClient client) {
    System.out.println("\n============================================================");
    System.out.println("🏗️ DOMAIN MODEL EXPORT EXAMPLES");
    System.out.println("============================================================");

    System.out.println("\n1️⃣ Searching for available domain models...");
    try {
      List<SimpleDomainSpecificModelPublicDto> domainModels = client.searchDomainModels();
      if (domainModels.isEmpty()) {
        System.out.println("🔭 No domain models found for export");
        return;
      }

      System.out.println("✅ Found " + domainModels.size() + " domain models:");
      for (int i = 0; i < Math.min(3, domainModels.size()); i++) {
        SimpleDomainSpecificModelPublicDto model = domainModels.get(i);
        System.out.println("   " + (i + 1) + ". " + model.getName() + " (" + model.getGuid() + ")");
      }

      SimpleDomainSpecificModelPublicDto selectedModel = domainModels.get(0);
      System.out.println("\n🎯 Using domain model: '" + selectedModel.getName() + "'");

      Map<String, Path> exportResults = new HashMap<>();

      System.out.println("\n2️⃣ Exporting domain model in multiple formats...");

      // PDF Export with auto-detection
      System.out.println("   📄 Exporting as PDF...");
      Optional<byte[]> pdfContent = client.exportDomainModelPdf(selectedModel.getGuid());
      if (pdfContent.isPresent()) {
        String baseFilename = "domain_model_pdf_" + selectedModel.getGuid();
        Optional<Path> pdfPath = exportWithDetection(pdfContent.get(), baseFilename, "pdf");
        if (pdfPath.isPresent()) {
          exportResults.put("PDF", pdfPath.get());
          System.out.println("   ✅ PDF exported: " + pdfPath.get());
        } else {
          System.out.println("   ❌ PDF export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ PDF export failed: No content received");
      }

      // OpenOffice Export with auto-detection
      System.out.println("   📝 Exporting as OpenOffice...");
      Optional<byte[]> odtContent = client.exportDomainModelOpenOffice(selectedModel.getGuid());
      if (odtContent.isPresent()) {
        String baseFilename = "domain_model_odt_" + selectedModel.getGuid();
        Optional<Path> odtPath = exportWithDetection(odtContent.get(), baseFilename, "odt");
        if (odtPath.isPresent()) {
          exportResults.put("OpenOffice", odtPath.get());
          System.out.println("   ✅ OpenOffice exported: " + odtPath.get());
        } else {
          System.out.println("   ❌ OpenOffice export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ OpenOffice export failed: No content received");
      }

      // OKSTRA Export with auto-detection
      System.out.println("   🏗️ Exporting as OKSTRA...");
      Optional<byte[]> okstraContent = client.exportDomainModelOkstra(selectedModel.getGuid());
      if (okstraContent.isPresent()) {
        String baseFilename = "domain_model_okstra_" + selectedModel.getGuid();
        Optional<Path> okstraPath = exportWithDetection(okstraContent.get(), baseFilename, "zip");
        if (okstraPath.isPresent()) {
          exportResults.put("OKSTRA", okstraPath.get());
          System.out.println("   ✅ OKSTRA exported: " + okstraPath.get());
        } else {
          System.out.println("   ❌ OKSTRA export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ OKSTRA export failed: No content received");
      }

      // LOIN-XML Export with auto-detection
      System.out.println("   🔗 Exporting as LOIN-XML...");
      Optional<byte[]> loinXmlContent = client.exportDomainModelLoinXml(selectedModel.getGuid());
      if (loinXmlContent.isPresent()) {
        String baseFilename = "domain_model_loin_" + selectedModel.getGuid();
        Optional<Path> xmlPath = exportWithDetection(loinXmlContent.get(), baseFilename, "zip");
        if (xmlPath.isPresent()) {
          exportResults.put("LOIN-XML", xmlPath.get());
          System.out.println("   ✅ LOIN-XML exported: " + xmlPath.get());
        } else {
          System.out.println("   ❌ LOIN-XML export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ LOIN-XML export failed: No content received");
      }

      // IDS Export with auto-detection
      System.out.println("   🆔 Exporting as IDS...");
      Optional<byte[]> idsContent = client.exportDomainModelIds(selectedModel.getGuid());
      if (idsContent.isPresent()) {
        String baseFilename = "domain_model_ids_" + selectedModel.getGuid();
        Optional<Path> idsPath = exportWithDetection(idsContent.get(), baseFilename, "xml");
        if (idsPath.isPresent()) {
          exportResults.put("IDS", idsPath.get());
          System.out.println("   ✅ IDS exported: " + idsPath.get());
        } else {
          System.out.println("   ❌ IDS export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ IDS export failed: No content received");
      }

      // Summary with file type information
      System.out.println("\n📈 Export Summary: " + exportResults.size() + "/5 formats successful");
      for (Map.Entry<String, Path> entry : exportResults.entrySet()) {
        String filename = entry.getValue().getFileName().toString();
        String extension = filename.substring(filename.lastIndexOf('.') + 1).toUpperCase();
        System.out.println("   ✅ " + entry.getKey() + " (" + extension + "): " + entry.getValue());
      }

      if (exportResults.size() < 3) {
        System.out.println(
            "💡 Note: Some export formats may require special permissions or project setup");
        System.out.println(
            "💡 Content type detection helps ensure correct file extensions are used");
      }

    } catch (Exception e) {
      logger.error("❌ Error in domain model export examples", e);
      System.err.println("❌ Error in domain model export examples: " + e.getMessage());
    }
  }

  /** Main method to run domain model export examples. */
  public static void main(String[] args) {
    System.out.println("======================================================================");
    System.out.println("🚀 BIM PORTAL DOMAIN MODEL EXPORT EXAMPLES");
    System.out.println("======================================================================");

    if (!checkCredentials()) {
      System.err.println("❌ Cannot run export examples without credentials");
      return;
    }

    if (!ExportUtils.isExportDirectoryWritable()) {
      System.err.println("❌ Export directory is not writable!");
      return;
    }

    System.out.println("🔧 Setting up authenticated client...");

    try {
      EnhancedBimPortalClient client = BimPortalClientBuilder.buildDefault();

      EnhancedBimPortalClient.HealthCheckResult healthCheck = client.performHealthCheck();
      System.out.println("🩺 Health check: " + healthCheck);

      if (!healthCheck.isApiAccessible()) {
        System.err.println("❌ API is not accessible. Please check configuration.");
        return;
      }

      runDomainModelExportExamples(client);

      System.out.println(
          "\n======================================================================");
      System.out.println("✅ DOMAIN MODEL EXPORT EXAMPLES COMPLETE!");
      System.out.println("======================================================================");
      System.out.println(
          "📂 Check the '" + BimPortalConfig.getExportPath() + "' directory for exported files");

      client.logout();
      System.out.println("👋 Logged out and session closed.");

    } catch (Exception e) {
      logger.error("❌ Error running domain model export examples", e);
      System.err.println("❌ Error running examples: " + e.getMessage());
    }
  }
}

File: Python_client/examples/export_examples/domain_model_export_example.py
"""
Domain-specific model export examples for BIM Portal Python client.

This module demonstrates domain-specific model export workflows in multiple formats including
PDF, OpenOffice, OKSTRA, LOIN-XML, and IDS with automatic content type detection.
"""

import os
import logging
from pathlib import Path
from typing import Dict, Optional

from dotenv import load_dotenv

from client.auth.auth_config import BIM_PORTAL_PASSWORD_ENV_VAR, BIM_PORTAL_USERNAME_ENV_VAR
from client.auth.auth_service_impl import AuthService
from client.enhanced_bim_client import EnhancedBimPortalClient
from client.config import BIMPortalConfig
from examples.export_examples.utils.common_utils import check_credentials
from examples.export_examples.utils.export_utils import ExportUtils

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Load environment variables
load_dotenv()



def run_domain_model_export_examples(client: EnhancedBimPortalClient):
    """
    Run domain-specific model export examples with comprehensive format support.

    Args:
        client: Enhanced BIM Portal client
    """
    print("\n" + "=" * 60)
    print("🗂️  DOMAIN MODEL EXPORT EXAMPLES")
    print("=" * 60)

    print("\n1️⃣ Searching for available domain models...")
    try:
        domain_models = client.search_domain_models()
        if not domain_models:
            print("🔭 No domain models found for export")
            return

        print(f"✅ Found {len(domain_models)} domain models:")
        for i, model in enumerate(domain_models[:3], 1):
            print(f"   {i}. {model.name} ({model.guid})")

        selected_model = domain_models[0]
        print(f"\n🎯 Using domain model: '{selected_model.name}'")

        export_results: Dict[str, Optional[Path]] = {}

        print("\n2️⃣ Exporting domain model in multiple formats...")

        # PDF Export with auto-detection
        print("   📄 Exporting as PDF...")
        pdf_content = client.export_domain_model_pdf(selected_model.guid)
        if pdf_content:
            base_filename = f"domain_model_pdf_{selected_model.guid}"
            pdf_path = ExportUtils.export_with_detection(pdf_content, base_filename, "pdf")
            if pdf_path:
                export_results['PDF'] = pdf_path
                print(f"   ✅ PDF exported: {pdf_path}")
            else:
                print("   ❌ PDF export failed: Could not save file")
                export_results['PDF'] = None
        else:
            print("   ❌ PDF export failed: No content received")
            export_results['PDF'] = None

        # OpenOffice Export with auto-detection
        print("   📝 Exporting as OpenOffice...")
        odt_content = client.export_domain_model_openoffice(selected_model.guid)
        if odt_content:
            base_filename = f"domain_model_odt_{selected_model.guid}"
            odt_path = ExportUtils.export_with_detection(odt_content, base_filename, "odt")
            if odt_path:
                export_results['OpenOffice'] = odt_path
                print(f"   ✅ OpenOffice exported: {odt_path}")
            else:
                print("   ❌ OpenOffice export failed: Could not save file")
                export_results['OpenOffice'] = None
        else:
            print("   ❌ OpenOffice export failed: No content received")
            export_results['OpenOffice'] = None

        # OKSTRA Export with auto-detection
        print("   🗂️ Exporting as OKSTRA...")
        okstra_content = client.export_domain_model_okstra(selected_model.guid)
        if okstra_content:
            base_filename = f"domain_model_okstra_{selected_model.guid}"
            okstra_path = ExportUtils.export_with_detection(okstra_content, base_filename, "zip")
            if okstra_path:
                export_results['OKSTRA'] = okstra_path
                print(f"   ✅ OKSTRA exported: {okstra_path}")
            else:
                print("   ❌ OKSTRA export failed: Could not save file")
                export_results['OKSTRA'] = None
        else:
            print("   ❌ OKSTRA export failed: No content received")
            export_results['OKSTRA'] = None

        # LOIN-XML Export with auto-detection
        print("   🔗 Exporting as LOIN-XML...")
        loin_xml_content = client.export_domain_model_loin_xml(selected_model.guid)
        if loin_xml_content:
            base_filename = f"domain_model_loin_{selected_model.guid}"
            xml_path = ExportUtils.export_with_detection(loin_xml_content, base_filename, "zip")
            if xml_path:
                export_results['LOIN-XML'] = xml_path
                print(f"   ✅ LOIN-XML exported: {xml_path}")
            else:
                print("   ❌ LOIN-XML export failed: Could not save file")
                export_results['LOIN-XML'] = None
        else:
            print("   ❌ LOIN-XML export failed: No content received")
            export_results['LOIN-XML'] = None

        # IDS Export with auto-detection
        print("   🆔 Exporting as IDS...")
        ids_content = client.export_domain_model_ids(selected_model.guid)
        if ids_content:
            base_filename = f"domain_model_ids_{selected_model.guid}"
            ids_path = ExportUtils.export_with_detection(ids_content, base_filename, "xml")
            if ids_path:
                export_results['IDS'] = ids_path
                print(f"   ✅ IDS exported: {ids_path}")
            else:
                print("   ❌ IDS export failed: Could not save file")
                export_results['IDS'] = None
        else:
            print("   ❌ IDS export failed: No content received")
            export_results['IDS'] = None

        # Summary with file type information
        successful_exports = len([path for path in export_results.values() if path is not None])
        total_formats = len(export_results)

        print(f"\n📈 Export Summary: {successful_exports}/{total_formats} formats successful")

        for format_name, path in export_results.items():
            if path:
                extension = path.suffix[1:].upper() if path.suffix else "UNKNOWN"
                print(f"   ✅ {format_name} ({extension}): {path}")
            else:
                print(f"   ❌ {format_name}: Export failed")

        if successful_exports < 3:
            print("💡 Note: Some export formats may require special permissions or project setup")
            print("💡 Content type detection helps ensure correct file extensions are used")

        # Additional domain model details
        print(f"\n📋 Domain Model Details:")
        try:
            detailed_model = client.get_domain_model(selected_model.guid)
            if detailed_model:
                print(f"   Name: {detailed_model.name}")
                print(f"   GUID: {detailed_model.guid}")
                if hasattr(detailed_model, 'description') and detailed_model.description:
                    print(f"   Description: {detailed_model.description}")
                if hasattr(detailed_model, 'version') and detailed_model.version:
                    print(f"   Version: {detailed_model.version}")
                if hasattr(detailed_model, 'discipline') and detailed_model.discipline:
                    print(f"   Discipline: {detailed_model.discipline}")
            else:
                print("   Could not retrieve detailed information")
        except Exception as e:
            print(f"   Error retrieving details: {e}")

    except Exception as e:
        logger.error(f"Error in domain model export examples: {e}")
        print(f"❌ Error in domain model export examples: {e}")


def main():
    """Main method to run domain model export examples."""
    print("=" * 70)
    print("🚀 BIM PORTAL DOMAIN MODEL EXPORT EXAMPLES")
    print("=" * 70)

    if not check_credentials():
        print("❌ Cannot run export examples without credentials")
        return

    if not ExportUtils.is_export_directory_writable():
        print("❌ Export directory is not writable!")
        return

    print("🔧 Setting up authenticated client...")

    try:
        auth_service = AuthService()
        client = EnhancedBimPortalClient(
            auth_service=auth_service,
            base_url=BIMPortalConfig.BASE_URL
        )

        # Run domain model export examples
        run_domain_model_export_examples(client)

        print("\n" + "=" * 70)
        print("✅ DOMAIN MODEL EXPORT EXAMPLES COMPLETE!")
        print("=" * 70)
        print(f"📂 Check the '{BIMPortalConfig.EXPORT_DIRECTORY}' directory for exported files")
        
    except Exception as e:
        logger.error(f"Error running domain model export examples: {e}")
        print(f"❌ Error running examples: {e}")


if __name__ == "__main__":
    main()